server/src/main/java/com/hedera/block/server/notifier/NotifierImpl.java

```diff
@@ -0,0 +1,158 @@
+ * producers.
+ *
+ * <p>As an implementation of the StreamMediator interface, it proxies block item persistence
+ * responses back to the producers as they arrive via a RingBuffer maintained in the base class and
+ * persists the block items to a store. It also notifies the mediator of critical system events and
+ * will stop the server in the event of an unrecoverable error.
+ */
+class NotifierImpl extends SubscriptionHandlerBase<PublishStreamResponse> implements Notifier {
+
+    private final System.Logger LOGGER = System.getLogger(getClass().getName());
+
+    private final Notifiable mediator;
+    private final MetricsService metricsService;
+    private final ServiceStatus serviceStatus;
+
+    NotifierImpl(
+            @NonNull
+                    final Map<
+                                    BlockNodeEventHandler<ObjectEvent<PublishStreamResponse>>,
+                                    BatchEventProcessor<ObjectEvent<PublishStreamResponse>>>
+                            subscribers,
+            @NonNull final Notifiable mediator,
+            @NonNull final BlockNodeContext blockNodeContext,
+            @NonNull final ServiceStatus serviceStatus) {
+
+        super(
+                subscribers,
+                blockNodeContext.metricsService().get(Producers),
+                blockNodeContext
+                        .configuration()
+                        .getConfigData(NotifierConfig.class)
+                        .ringBufferSize());
+
+        this.mediator = mediator;
+        this.metricsService = blockNodeContext.metricsService();
+        this.serviceStatus = serviceStatus;
+    }
+
+    @Override
+    public void notifyUnrecoverableError() {
+
+        mediator.notifyUnrecoverableError();
+
+        // Publish an end of stream response to the producers.
+        final PublishStreamResponse errorStreamResponse = buildErrorStreamResponse();
+        ringBuffer.publishEvent((event, sequence) -> event.set(errorStreamResponse));
+
+        // Stop the server
+        serviceStatus.stopWebServer(getClass().getName());
+    }
+
+    /**
+     * Publishes the given block item to all subscribed producers.
+     *
+     * @param blockItem the block item from the persistence layer to publish a response to upstream
+     *     producers
+     */
+    @Override
+    public void publish(@NonNull BlockItem blockItem) {
+
+        try {
+            if (serviceStatus.isRunning()) {
+                // Publish the block item to the subscribers
+                final var publishStreamResponse =
+                        PublishStreamResponse.newBuilder()
+                                .acknowledgement(buildAck(blockItem))
+                                .build();
+                ringBuffer.publishEvent((event, sequence) -> event.set(publishStreamResponse));
+
+                metricsService.get(SuccessfulPubStreamResp).increment();
+            } else {
+                LOGGER.log(ERROR, "Notifier is not running.");
```

Comment:
Shouldn't we throw error here as well? Can say for sure, just wondering if just logging is sufficient to handle this scenario.

Reply:
`serviceStatus` is a singleton shared by all of the main components. With tens of thousands of block items transiting the system at any given point of time, the idea was to quickly stop processing everywhere asap when there's an exception. In this particular part of the code, the else case is triggered when `serviceStatus.isRunning() == false`. That flag was switched from an exception somewhere else in the system (for example, in `ProducerBlockItemObserver`) so here we just need to stop publishing responses back to the producers. I think throwing an exception at this point in the code would inject stacktraces in the logs that might obscure where the actual exception happened.

Reply:
Gotcha, thank you for the answer 👍